### PR TITLE
Using wildcards in Data Sampling

### PR DESCRIPTION
**Pull request by knopers8**

### CODE REVIEW

**Comment by knopers8** [comment]
@Barthelemy This is ready for review. It will break `o2-qc-run-advanced`, but the rest of QC should remain intact. For a moment, Data Sampling can support both the old format of specyfing inputs and the new - queries, so there is nothing significant broken.

File: Framework/Core/include/Framework/DataSamplingPolicy.h [hunk 1]
```diff
@@ -36,39 +37,19 @@ namespace framework
 class DataSamplingPolicy
 {
  private:
-  // todo: see if dpl matchers can be used here instead of this strange construction
-  struct inputSpecHasher {
-    size_t operator()(const InputSpec& i) const
+  using V = std::vector<std::pair<InputSpec, OutputSpec>>;
```
Comment (Barthelemy) [inline Framework/Core/include/Framework/DataSamplingPolicy.h]: CodingGuidelines specify that names should be "meaningful; abbreviations should be avoided" . Do you think that there is a more meaningful name for `V` ? if not I will let it go.

File: Framework/Core/src/DataSamplingPolicy.cxx [hunk 2]
```diff
@@ -42,34 +43,58 @@ void DataSamplingPolicy::configure(const ptree& config)
...
+  auto subSpec = subSpecString.find_first_of("-*") != std::string::npos ? -1 : std::strtoull(subSpecString.c_str(), nullptr, 10);
+
   mPaths.clear();
   size_t outputId = 0;
+  std::vector<InputSpec> inputSpecs;
+
+  if (config.get_optional<std::string>("query").has_value()) {
+    inputSpecs = DataDescriptorQueryBuilder::parse(config.get<std::string>("query").c_str());
+  } else {
+    // for a while, we leave an old way of specyfing the inputs, so we can gracefully update QC
```
Comment (Barthelemy) [inline Framework/Core/src/DataSamplingPolicy.cxx]: thank you for that !

File: Framework/Core/test/test_DataSampling.cxx [hunk 3]
```diff
@@ -43,8 +43,8 @@ BOOST_AUTO_TEST_CASE(DataSamplingSimpleFlow)
 
   std::string configFilePath = std::string(getenv("O2_ROOT")) + "/share/tests/test_DataSampling.json";
   std::cout << "config file : "
-            << "json://" << configFilePath << std::endl;
-  DataSampling::GenerateInfrastructure(workflow, "json://" + configFilePath);
+            << "json:/" << configFilePath << std::endl;
+  DataSampling::GenerateInfrastructure(workflow, "json:/" + configFilePath);
```
Comment (Barthelemy) [inline Framework/Core/test/test_DataSampling.cxx]: why is this change needed ?

File: Framework/Core/test/test_DataSamplingPolicy.cxx [hunk 4]
```diff
@@ -59,19 +59,7 @@ BOOST_AUTO_TEST_CASE(DataSamplingPolicyConfiguration)
...
-  dh1.put("dataDescription", "CHLEB");
-  ptree dh2;
-  dh2.put("binding", "");
-  dh2.put("dataOrigin", "TST");
-  dh2.put("dataDescription", "MLEKO");
-  dataHeaders.push_back(std::make_pair("", dh1));
-  dataHeaders.push_back(std::make_pair("", dh2));
-  config.add_child("dataHeaders", dataHeaders);
-  config.put("subSpec", "33");
+  config.put("query", "c:TST/CHLEB/33;m:TST/MLEKO/33");
```
Comment (Barthelemy) [inline Framework/Core/test/test_DataSamplingPolicy.cxx]: a nice example of how much neat the syntax became !

**Comment by Barthelemy** [comment]
See my comments, otherwise it seems very good !

File: Framework/Core/include/Framework/DataSamplingPolicy.h (hunk 1 above)
Comment (knopers8) [reply Framework/Core/include/Framework/DataSamplingPolicy.h]: Hmm, maybe `PathVectorBase`?

File: Framework/Core/test/test_DataSampling.cxx (hunk 3 above)
Comment (knopers8) [reply Framework/Core/test/test_DataSampling.cxx]: I noticed that Configuration, when it couldn't find a file, was printing the path with double `/` at the beginning, like so:
```
Unable to read JSON file: //home/pkonopka/alice/sw/slc7_x86-64/QualityControl/use-workflow-merging-1/etc/basicaaa.json
```
So as I understand, the preferred or default way is to use one `/` in `json:/`. That being said, both alternatives work when there is a file available at specified path.

File: Framework/Core/include/Framework/DataSamplingPolicy.h (hunk 1 above)
Comment (Barthelemy) [reply Framework/Core/include/Framework/DataSamplingPolicy.h]: ok

File: Framework/Core/test/test_DataSampling.cxx (hunk 3 above)
Comment (Barthelemy) [reply Framework/Core/test/test_DataSampling.cxx]: ok

**Comment by Barthelemy** [comment]
For me that can be merged once you cahnge the name of `V`.

File: Framework/Core/include/Framework/DataSamplingPolicy.h (hunk 1 above)
Comment (knopers8) [reply Framework/Core/include/Framework/DataSamplingPolicy.h]: done

**Comment by knopers8** [comment]
Actually, I still need to update the documentation, hold your horses

**Comment by knopers8** [comment]
Ok, now it is ready.

**Comment by Barthelemy** [comment]
If the tests fail for no reason we will ask @awegrzyn to merge.

**Comment by knopers8** [comment]
`build/O2/o2` seems stuck.
